Example: replicated deployments (#1632)
liqotech/liqo: docs/examples/offloading-multiple-pods.md

> # Offloading multiple pods

Reviewer: Offloading pods on multiple remote clusters at once

> # Offloading multiple pods

Reviewer: Additional comment: I would update the image with the names of the clusters (e.g., destination cluster #1, etc.)

> # Offloading multiple pods

Author: Maybe just **Offloading on multiple remote clusters** would be better to avoid double lines in the documentation menù?

![image](https://user-images.githubusercontent.com/33266330/214828703-a6354598-42b6-47af-8f8a-c16853463ff4.png)

![image](https://user-images.githubusercontent.com/33266330/214828813-0c6addec-66ac-4157-9091-548fa7c98152.png)

> In this tutorial you will learn how to deploy an application, and use liqo to replicate it on other clusters.

Reviewer: ```suggestion
In this tutorial you will learn how to deploy an application, and use Liqo to replicate it on multiple clusters.
```

![Offloading Multiple Pods Overview](/_static/images/examples/offloading-multiple-pods/offloading-multiple-pods.drawio.svg)

> More specifically, you will configure a scenario composed of a *single entry point cluster* leveraged for the deployment of the applications (called *origin cluster*) and two *destination clusters*.

Reviewer: ```suggestion
In this example you will configure a scenario composed of a *single entry point cluster* leveraged for the deployment of the applications (called *origin cluster*) and two *destination clusters*.
```

> The deployed application will be replicated on every *destination cluster* in order to deploy exactly **one** identical *pod* on each *destination cluster*.

Reviewer: ```suggestion
The deployed application will be replicated on all *destination clusters* in order to deploy exactly **one** identical *pod* on each *destination cluster*.
```

## Provision the playground

> First, check that you are compliant with the [requirements](/examples/requirements.md).

Reviewer: ```suggestion
First, make sure that the [requirements](/examples/requirements.md) for Liqo are satisfied.
```

Then, let's open a terminal on your machine and launch the following script, which creates the three above-mentioned clusters with KinD and installs Liqo on all of them.

{{ env.config.html_context.generate_clone_example('offloading-multiple-pods') }}

Export the kubeconfigs environment variables to use them in the rest of the tutorial:

```bash
export KUBECONFIG=liqo_kubeconf_origin
export KUBECONFIG_DESTINATION1=liqo_kubeconf_destination1

> export KUBECONFIG_DESTINATION2=liqo_kubeconf_destination2

Reviewer 2: The other examples have cities' names for clusters. I would follow this pattern.

They can be something like `europe-cloud`, `europe-yourcity1-edge`, and `europe-yourcity2-edge` (choose 2 cities you like).

```

```{admonition} Note
We suggest exporting the kubeconfig of the *origin* cluster as default (i.e., `KUBECONFIG`), since you will mainly interact with it.
```

> At this point, you should have three clusters with Liqo installed on them.

Reviewer: ```suggestion
Now you should have three clusters with Liqo running.
```

The setup script named them **origin**, **destination1** and **destination2**, and respectively configured the following cluster labels:

* *origin*: `topology.liqo.io/type=origin`
* *destination1*: `topology.liqo.io/type=destination`
* *destination2*: `topology.liqo.io/type=destination`

You can check that the clusters are correctly labeled through:

```bash
liqoctl status
liqoctl status --kubeconfig "$KUBECONFIG_DESTINATION1"
liqoctl status --kubeconfig "$KUBECONFIG_DESTINATION1"
```

## Peer the clusters

> Once Liqo is installed in your clusters, you can establish new *peerings*.

Reviewer: ```suggestion
Now, you can establish new Liqo *peerings* from *origin* to *destination* clusters, e.g., using the [out-of-band peering](FeaturesPeeringOutOfBandControlPlane) approach:
```
(I would delete the next two sentences, which are summarized here)

> In this example, you will use the [in-band peering approach](FeaturesPeeringInBandControlPlane).

Reviewer 2: Please, use the `out-of-band` since `in-band` is not required

To establish the peerings from *origin* to *destination* clusters run:

```bash
liqoctl peer in-band --remote-kubeconfig "$KUBECONFIG_DESTINATION1"
liqoctl peer in-band --remote-kubeconfig "$KUBECONFIG_DESTINATION2"
```

When the above commands return successfully, you can check the peering status by running:

```bash
kubectl get foreignclusters
```

> The output should look like the following, indicating that an outgoing peering is currently active towards both the *destination1* and the *destination2* clusters, as well as the cross-cluster network tunnels have been established:

Reviewer: ```suggestion
The output should look like the following, indicating that an outgoing peering is currently active towards both the *destination1* and the *destination2* clusters, with the two cross-cluster network tunnels marked as `established`:
```

```text
NAME           TYPE     OUTGOING PEERING   INCOMING PEERING   NETWORKING    AUTHENTICATION   AGE
destination1   InBand   Established        None               Established   Established      41s
destination2   InBand   Established        None               Established   Established      7s
```

Additionally, you should have two new virtual nodes in the *origin* cluster, characterized by the install-time provided labels:

```bash
kubectl get node --selector=liqo.io/type=virtual-node --show-labels
```

```text
NAME            STATUS   ROLES   AGE     VERSION   LABELS
NAME                STATUS   ROLES   AGE   VERSION   LABELS
liqo-destination1   Ready    agent   57s   v1.25.0   liqo.io/remote-cluster-id=df0baf71-c428-4d2e-8c97-3559e205d34a,liqo.io/type=virtual-node,topology.liqo.io/type=destination
liqo-destination2   Ready    agent   43s   v1.25.0   liqo.io/remote-cluster-id=3c4604aa-b9b4-4442-be6d-8ffc713139fc,liqo.io/type=virtual-node,topology.liqo.io/type=destination
```

```{admonition} Note
Some of the default labels were omitted for the sake of clarity.
```

## Tune namespace offloading

> Now, let's suppose you want to deploy an application that needs to be scheduled on every *destination* cluster, but not in the *origin* one.

Reviewer: ```suggestion
Now, let's pretend you want to deploy an application that needs to be scheduled on all *destination* clusters, but not in the *origin* one.
First, we create a new namespace, then enable Liqo offloading to it:
```
(move both commands in the same 'bash' section)

```bash
kubectl create namespace liqo-demo
```

Then, enable Liqo offloading for that namespace:

```bash
liqoctl offload namespace liqo-demo \
  --namespace-mapping-strategy EnforceSameName \
  --pod-offloading-strategy Remote \

>   --selector 'topology.liqo.io/type=destination'

Reviewer 2: This is useless: you are selecting all clusters

Reviewer: I would keep it anyway as an example, e.g., to show how to deal with the case in which you have two types of edge clusters (edge-italy, edge-france).

Author: I know is useless , but I did it to show how to manage a case in which not all remote clusters have to be considered for the offloading. I could:
1.  add a **note** to explain that and keeps the command as it is
2.  add a **note** to explain that is possible to add the flag --selector to specify the destination clusters and remove the --selector flag from the command.

Reviewer 2: In my opinion, the second is better. You can say something like: "You can select a subset of remote clusters by using the `--selector` flag (e.g., selecting all clusters in the region `edge-italy`. For further details, try https://docs.liqo.io/en/v0.6.1/examples/offloading-with-policies.html#tune-namespace-offloading"

```

> The above command configures the following aspects (see the dedicated [usage page](/usage/namespace-offloading.md) for additional information concerning namespace offloading configurations):

Reviewer: ```suggestion
The above command configures Liqo with the following behaviour (see the dedicated [usage page](/usage/namespace-offloading.md) for additional information concerning namespace offloading configurations):
```

> * the `liqo-demo` namespace, and the contained resources, are offloaded only to the clusters with the `topology.liqo.io/type=destination` label.

Reviewer 2: I would replace it with a box saying something like "Pro Tip: you can select a subset of your cluster as explained in the [Offloading with Policies](https://docs.liqo.io/en/v0.6.1/examples/offloading-with-policies.html) demo"

* the pods living in the `liqo-demo` namespace only on virtual nodes.

> The *NamespaceOffloading* resource created by *liqoctl* in the `liqo-demo` namespace exposes the status of the offloading process, including a global *OffloadingPhase*, which is expected to be `Ready`, and a list of *RemoteNamespaceConditions*, one for each remote cluster.

Reviewer 2: This should be explained in the other example, if not add it in that page

In this case both the *destination1* and *destination2* clusters have been selected to offload the namespace `liqo-demo`.

```bash
kubectl get namespaceoffloadings offloading -n liqo-demo -o yaml
```

> 

Reviewer: I would add something which explains what is the previous command about, and that the next block of text represents an example of a possible output.

Author: This part has been removed and a link to another examples which explains this feature has been added (to avoid duplicated information).

      message: The remote cluster has been selected through the ClusterSelector
        field
      reason: ClusterSelected
      status: "True"
      type: OffloadingRequired
    - lastTransitionTime: "2023-01-22T21:26:11Z"
      message: Namespace correctly offloaded to the remote cluster
      reason: NamespaceCreated
      status: "True"
      type: Ready
    destination2-9ea373:
    - lastTransitionTime: "2023-01-22T21:26:11Z"
      message: The remote cluster has been selected through the ClusterSelector
        field
      reason: ClusterSelected
      status: "True"
      type: OffloadingRequired
    - lastTransitionTime: "2023-01-22T21:26:11Z"
      message: Namespace correctly offloaded to the remote cluster
      reason: NamespaceCreated
      status: "True"
      type: Ready
```

Indeed, if you query for the namespaces in the *destination1* or *destination2* cluster, you should see the following output, confirming that the remote namespace has been correctly created by Liqo:

```bash
kubectl get namespaces liqo-demo --kubeconfig="$KUBECONFIG_DESTINATION1"
kubectl get namespaces liqo-demo --kubeconfig="$KUBECONFIG_DESTINATION2"
```

> Both the commands should return the following output:

Reviewer: ```suggestion
Both the commands should return an output similar to the following:
```

```text
NAME        STATUS   AGE
liqo-demo   Active   70s
```

## Deploy applications

Now it's time to deploy the application.

> In order to create a replica of the application in every *destination* cluster is necessary to enforce these conditions.

Reviewer: ```suggestion
In order to create a replica of the application in each *destination* cluster, you need to enforce the following conditions:
```
(move the next text as bullet list)

* The *deployment* resource must produce at least one *pod* for each *destination* cluster.
* Each *destination* cluster must schedule at most one *pod* on its nodes.

> To get this result you can exploit these *kubernetes* features:

Reviewer: ```suggestion
To obtain this result you can leverage the following features available in *kubernetes*:
```
(move the following text as a bullet list)

* Set a number of replicas in the *deployment* which is equal to the number of *destination* clusters

> * Set *topologySpreadConstraints* inside the *deployment*'s template, which sets the *maxSkew* equal to **1**.

Reviewer 2: Add a link to the Kubernetes documentation for this feature

> The file `manifests/deployment.yaml` contains an example of a *deployment* which satisfies these conditions.

Reviewer: Not clear to me. It is `manifests/deployment.yaml` a working link?

Author: No, is a file contained inside the example folder, I added a `./` in front of the path to make it more clear and compliant with the rest of the documentation.

Let's deploy it:

```bash
kubectl apply -f ./manifests/deployment.yaml -n liqo-demo
```

```{admonition} More replicas

> If the deployment uses a number of replicas which is higher than the number of *virtual nodes*, the pods will be scheduled respecting the *maxSkew* value, which guarantees that the difference between the maximum number of pods (scheduled on a single node) and the minimum will be **1**

Reviewer: ```suggestion
If the deployment uses a number of replicas which is higher than the number of *virtual nodes*, the pods will be scheduled respecting the *maxSkew* value, which guarantees that the difference between the maximum number of pods (scheduled on a single node) and the minimum will be **1**.
```
(honestly, I did not understand exactly what happens, but I have to say that is likely a *my* problem)

Author: I added a link to documentation about *maxSkew* to make it more clear. It is a bit tricky to understand.

```

> Checking the pod status, it is possible to verify that one has been scheduled onto the *destination1* cluster, and the other on *destination2*, and it is correctly running.

Reviewer: ```suggestion
We can check the pod status and verify that each *destination* cluster has scheduled one *pod* on its nodes, i.e., one pod has been scheduled onto the *destination1* cluster, and the other on *destination2*, and they are both correctly running:
```
(I would remove next sentence, which has been inglobated in the one above)

Every *destination* cluster has scheduled one *pod* on its nodes.

```bash
kubectl get pod -n liqo-demo -o wide
```

```text
NAME                            READY   STATUS    RESTARTS   AGE     IP            NODE                NOMINATED NODE   READINESS GATES
liqo-demo-app-777fb9fc8-bbt4d   1/1     Running   0          7m28s   10.113.0.65   liqo-destination1   <none>           <none>
liqo-demo-app-777fb9fc8-wrjph   1/1     Running   0          7m28s   10.109.0.62   liqo-destination2   <none>           <none>
```

## Tear down the playground

> 

Reviewer: ```suggestion
Our example is finished; now we can remove all the created resources and tear down the playground.
```

### Unoffload namespaces

Before starting the uninstallation process, make sure that all namespaces are unoffloaded:

```bash
liqoctl unoffload namespace liqo-demo
```

### Revoke peerings

Similarly, make sure that all the peerings are revoked:

```bash
liqoctl unpeer in-band --remote-kubeconfig "$KUBECONFIG_DESTINATION1" 
liqoctl unpeer in-band --remote-kubeconfig "$KUBECONFIG_DESTINATION2" 
```

At the end of the process, the virtual nodes are removed from the local cluster.

### Uninstall Liqo

Now you can uninstall Liqo from your clusters:

```bash
liqoctl uninstall
liqoctl uninstall --kubeconfig="$KUBECONFIG_DESTINATION1" 
liqoctl uninstall --kubeconfig="$KUBECONFIG_DESTINATION2" 
```

> ```{admonition} Purge

Reviewer: I would avoid this detail about 'purge' repeated in each example.
Can we move this note into a more general place, and add a link to that page here?
(e.g., Detailed instructions about how to remove Liqo are available in THIS page).

Author: It could be done in another PR. In my opinion, the command examples inside the **admonition** could be removed from this example.
